src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs

```diff
@@ -1064,67 +1069,68 @@ public SolutionCompilationState WithOptions(SolutionOptionSet options)
                 // linked document ids our state says are in it.  Note: all of these trees should share the same green
                 // trees, as that is setup in Soltion.WithFrozenPartialCompilationIncludingSpecificDocument.  This helps
                 // ensure that the cost here is low for files with lots of linked siblings.
                 foreach (var currentDocumentId in allDocumentIds)
                 {
-                    var document = @this.SolutionState.GetRequiredDocumentState(currentDocumentId);
-                    builder.Add((document, document.GetSyntaxTree(cancellationToken)));
+                    var documentState = @this.SolutionState.GetRequiredDocumentState(currentDocumentId);
+                    statesAndTrees.Add((documentState, documentState.GetSyntaxTree(cancellationToken)));
                 }
 
                 using (@this.StateLock.DisposableWait(cancellationToken))
                 {
-                    SolutionCompilationState? currentPartialSolution = null;
-                    @this._latestSolutionWithPartialCompilation?.TryGetTarget(out currentPartialSolution);
-
-                    var reuseExistingPartialSolution =
-                        (DateTime.UtcNow - @this._timeOfLatestSolutionWithPartialCompilation).TotalSeconds < 0.1 &&
-                        @this._documentIdOfLatestSolutionWithPartialCompilation == documentId;
-
-                    if (reuseExistingPartialSolution && currentPartialSolution != null)
+                    if (@this._cachedFrozenDocumentState.TryGetValue(documentId, out var compilationState))
                     {
                         SolutionLogger.UseExistingPartialSolution();
-                        return currentPartialSolution;
                     }
-
-                    var newIdToProjectStateMap = @this.SolutionState.ProjectStates;
-                    var newIdToTrackerMap = @this._projectIdToTrackerMap;
-
-                    foreach (var (doc, tree) in builder)
+                    else
                     {
-                        // if we don't have one or it is stale, create a new partial solution
-                        var tracker = @this.GetCompilationTracker(doc.Id.ProjectId);
-                        var newTracker = tracker.FreezePartialStateWithTree(@this, doc, tree, cancellationToken);
-
-                        Contract.ThrowIfFalse(newIdToProjectStateMap.ContainsKey(doc.Id.ProjectId));
-                        newIdToProjectStateMap = newIdToProjectStateMap.SetItem(doc.Id.ProjectId, newTracker.ProjectState);
-                        newIdToTrackerMap = newIdToTrackerMap.SetItem(doc.Id.ProjectId, newTracker);
+                        compilationState = ComputeFrozenPartialState(@this, statesAndTrees, cancellationToken);
+                        @this._cachedFrozenDocumentState.Add(documentId, compilationState);
+                        SolutionLogger.CreatePartialSolution();
```

Comment:
One thing that might be useful here would be to add telemetry for how big that dictionary is getting; the way this change could break something would be if previously we had a bunch of different requests for freezing different documents, and now we're creating a lot of different snapshots and holding onto all of them. I can't imagine a way that'd happen. But maybe consider adding a quick check like "if the count is more than 5, trigger a fault."

If you want to delete SolutionLogger I wouldn't mind though I'm not sure if we've really ever looked at that?

Reply:
> If you want to delete SolutionLogger I wouldn't mind though I'm not sure if we've really ever looked at that?

WFM.